@*
 * Copyright 2010-2013 Artima, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *@

@userGuidePage("Using assertions") {
<div style="text-align: left">

<h1>Using assertions</h1>

<p>
ScalaTest makes three assertions available by default in any style trait. You can use:
</p>

<ul>
<li><code>assert</code> with <code>===</code>, for general assertions;</li>
<li><code>expect</code>, to differentiate expected from actual values;</li>
<li><code>intercept</code>, to ensure a bit of code throws an expected exception.</li>
</ul>

<p>
To get moving quickly in ScalaTest, learn and use these three assertions. Later if you prefer
you can switch to the more expressive <a href="/user_guide/using_matchers">matchers DSL</a>.
</p>

<p>
In any Scala program, you can write assertions by invoking <code>assert</code> and passing in a <code>Boolean</code> expression,
such as:
</p>

<pre class="stHighlighted">
<span class="stReserved">val</span> left = <span class="stLiteral">2</span>
<span class="stReserved">val</span> right = <span class="stLiteral">1</span>
assert(left == right)
</pre>

<p>
If the passed expression is <code>true</code>, <code>assert</code> will return normally. If <code>false</code>,
<code>assert</code> will complete abruptly with an <code>AssertionError</code>. This behavior is provided by
the <code>assert</code> method defined in object <code>Predef</code>, whose members are implicitly imported into every
Scala source file.
</p>

<p>
ScalaTest defines another <code>assert</code> method that hides the
one in <code>Predef</code>. It behaves the same, except that if <code>false</code> is passed it throws
<code>TestFailedException</code> instead of <code>AssertionError</code>. The reason it throws <code>TestFailedException</code>
is because <code>TestFailedException</code> carries information about exactly which item in the stack trace represents
the line of test code that failed, which can help users more quickly find an offending line of code in a failing test.
<p>

<p>
If you pass the previous <code>Boolean</code> expression, <code>left == right</code> to <code>assert</code> in a ScalaTest test, a failure
will be reported, but without reporting the left and right values. You can alternatively encode these values in a <code>String</code> passed as
a second argument to <code>assert</code>, like this:
</p>

<pre class="stHighlighted">
<span class="stReserved">val</span> left = <span class="stLiteral">2</span>
<span class="stReserved">val</span> right = <span class="stLiteral">1</span>
assert(left == right, left + <span class="stQuotedString">" did not equal "</span> + right)
</pre>

<p>
Using this form of <code>assert</code>, the failure report will include the left and right values, thereby
helping you debug the problem. However, ScalaTest provides the <code>===</code> operator to make this easier.
You use it like this:
</p>

<pre class="stHighlighted">
<span class="stReserved">val</span> left = <span class="stLiteral">2</span>
<span class="stReserved">val</span> right = <span class="stLiteral">1</span>
assert(left === right)
</pre>

<p>
Because you use <code>===</code> here instead of <code>==</code>, the failure report will include the left
and right values. For example, the detail message in the thrown <code>TestFailedException</code> from the <code>assert</code>
shown previously will include, "2 did not equal 1".
From this message you will know that the operand on the left had the value 2, and the operand on the right had the value 1.
</p>

<h2>Expected results</h2>

Although <code>===</code> provides a natural, readable extension to Scala's <code>assert</code> mechanism,
as the operands become lengthy, the code becomes less readable. In addition, the <code>===</code> comparison
doesn't distinguish between actual and expected values. The operands are just called <code>left</code> and <code>right</code>,
because if one were named <code>expected</code> and the other <code>actual</code>, it would be difficult for people to
remember which was which. To help with these limitations of <code>assert</code>, ScalaTest provides a method called <code>expect</code> that
can be used as an alternative to <code>assert</code> with <code>===</code>. To use <code>expect</code>, you place
the expected value in parentheses after <code>expect</code>, followed by curly braces containing code 
that should result in the expected value. For example:

<pre class="stHighlighted">
<span class="stReserved">val</span> a = <span class="stLiteral">5</span>
<span class="stReserved">val</span> b = <span class="stLiteral">2</span>
expect(<span class="stLiteral">2</span>) {
  a - b
}
</pre>

<p>
In this case, the expected value is <code>2</code>, and the code being tested is <code>a - b</code>. This expectation will fail, and
the detail message in the <code>TestFailedException</code> will read, "Expected 2, but got 3."
</p>

<h2>Intercepted exceptions</h2>

<p>
Sometimes you need to test whether a method throws an expected exception under certain circumstances, such
as when invalid arguments are passed to the method. You can do this in the JUnit 3 style, like this:
</p>

<pre class="stHighlighted">
<span class="stReserved">val</span> s = <span class="stQuotedString">"hi"</span>
<span class="stReserved">try</span> {
  s.charAt(-<span class="stLiteral">1</span>)
  fail()
}
<span class="stReserved">catch</span> {
  <span class="stReserved">case</span> _: <span class="stType">IndexOutOfBoundsException</span> => <span class="stLineComment">// Expected, so continue</span>
}
</pre>

<p>
If <code>charAt</code> throws <code>IndexOutOfBoundsException</code> as expected, control will transfer
to the catch case, which does nothing. If, however, <code>charAt</code> fails to throw an exception,
the next statement, <code>fail()</code>, will be run. The <code>fail</code> method always completes abruptly with
a <code>TestFailedException</code>, thereby signaling a failed test.
</p>

<p>
To make this common use case easier to express and read, ScalaTest provides an <code>intercept</code>
method. You use it like this:
</p>

<pre class="stHighlighted">
<span class="stReserved">val</span> s = <span class="stQuotedString">"hi"</span>
intercept[<span class="stType">IndexOutOfBoundsException</span>] {
  s.charAt(-<span class="stLiteral">1</span>)
}
</pre>

<p>
This code behaves much like the previous example. If <code>charAt</code> throws an instance of <code>IndexOutOfBoundsException</code>,
<code>intercept</code> will return that exception. But if <code>charAt</code> completes normally, or throws a different
exception, <code>intercept</code> will complete abruptly with a <code>TestFailedException</code>. <code>intercept</code> returns the
caught exception so that you can inspect it further if you wish, for example, to ensure that data contained inside
the exception has the expected values. Here's an example:
</p>

<pre class="stHighlighted">
<span class="stReserved">val</span> s = <span class="stQuotedString">"hi"</span>
val thrown = intercept[<span class="stType">IndexOutOfBoundsException</span>] {
  s.charAt(-<span class="stLiteral">1</span>)
}
assert(thrown.getMessage === <span class="stLiteral">"String index out of range: -1"</span>)
</pre>

<p>
Next, learn about <a href="tagging_your_tests">tagging your tests</a>.
</p>

</div>
}
